Testio app implementation

jncore opened the pull request:
Good day,

Here is my implementation. Had lots of fun writing it :)

Cheers.

Mindaugas-Jacionis commented:
Hi! I am glad that you had fun implementing the solution, it sure was interesting to review it. I really liked that you used `styled-components`, also you covered tests part real good. I have few questions/requests as an outcome of the review, would be great if you could find some time for them. :)

- What was the reason you ejected your `create-react-app`?
- FYI: in your `package.json` scripts you can do `eslint ./src` for linting command
- How would you handle linting when there is config files at the top level dir that need to be linted as well and some files that need to be ignored?
- What would be your approach to ensure quality of the commits? 
- Now you are using `react-icons` what would be your approach with custom icons, as now you are forced to use the ones provided by package.
- What would be your approach toward ui components that are shared across multiple routes/pages? 
- Extending on previous, what would be your approach towards shared brand style/functionality across multiple SPAs?
- What was the reason to choose sagas? Do you know any alternatives? What was benefits of using sagas instead of other options?
- I can see you are doing binding in constructor? What are other binding options and why you chose current one?
- You are using PrivateRoute pattern for routing? What is the reason to `checkAuthToken` in Servers screen? PrivateRoute is supposed to eliminate this need for explicit checking.
- Is there any specific reasoning why you chose to implement functional components as const arrow functions and not just simple functions?
- What would be your approach towards app that has huge amount of available api calls/actions? In current setup each axios request is written separately? Also in current app you would have(if there is 50-100 available API calls) very bloated `actionsTypes.js` - how would you solve that?
- I can see you are using named imports for your lodash functions? Could you explain why it is better over `import _ from 'lodash'` and then calling function `_.[functionName]()`? 
- In current app you need to persist only token? Lets assume you have big SPA where some data needs to be persisted and some not? How would you approach that? Also, how would you approach avoiding unnecessary http requests when needed data is already in storage? Or what would be your approach towards persisting data with "time to live"(i.e. you request user billing history and keep it in storage, but it is considered to be outdated every 30minutes)?
- How would you implement loading indicator for login/servers/in general any screen whit current setup of your app?
- Could you extend your PrivateRoute so it would be possible to pass in component that will be shown when auth requirements are not met?
- Lets assume you have more than one PrivateRoute and user is trying to access one of them without auth token. User is redirected to login and from there to servers(which was not the one that user was trying to access via direct link), how would solve such case when after login user has to land to that PrivateRoute which he/she intended to visit in the first place?
- I can see babel config is done in `package.json`(this is the result of ejected create-react-app) - what is your thoughts on such approach? Any other possible implementation of babel config?

jncore commented:
Hey, thanks for the feedback and some awesome questions! :)
Sorry for the delay, now I've committed my changes, and here are my thoughts.

----------
> What was the reason you ejected your create-react-app?

Looking back at it, I think it was not really necessary in this particular case. I did it in order to setup things like Enzyme adapter and localStorage mock for testing purposes, modify the ESLint configuration, and have more control over the environment. But I didn’t end up configuring much, so next time I plan to not eject for as long as I can, unless I need to write some middle-end/back-end layer with Node.js, or make drastic changes to the setup. :)

----------
> FYI: in your package.json scripts you can do eslint ./src for linting command

Thanks, I missed it :)

----------
> How would you handle linting when there is config files at the top level dir that need to be linted as well and some files that need to be ignored?

I would setup the ```.eslintignore``` config in the root folder, and specify the ignore rules for files and directories that I need to skip. I’ll setup a small example of that to showcase.

----------
> What would be your approach to ensure quality of the commits?

First of all, I’d setup an enforced pre-commit hook, which runs tests, ESLint, and maybe a tool like Prettier. But that doesn’t ensure the quality of commits automatically, so I would also collaborate with the team as much as I can to come up with a list of agreements on the way we handle our commits in the future. This includes things like consistent commit naming, separation of features between commits, and so on.

----------
> Now you are using react-icons what would be your approach with custom icons, as now you are forced to use the ones provided by package.

I’d wrap the custom SVG icons into React components, and put them into a separate package, so I would be able to import and reuse them in multiple SPAs.

----------
> What would be your approach toward ui components that are shared across multiple routes/pages?

In a current setup I would organize them in the ```shared``` or ```global``` folder inside the ```src/components```, and reuse them. Also, I’d advocate to keep the reusable UI components stateless and as simple as possible.

----------
> Extending on previous, what would be your approach towards shared brand style/functionality across multiple SPAs?

I would create a separate repository to host all shared components/styles. And then I’d just import the package into SPAs. If the app grows, the shared repository can be split up into separate functional/contextual repos containing components that belong together, to control the mess and complexity.

----------
> What was the reason to choose sagas? Do you know any alternatives? What was benefits of using sagas instead of other options?

I chose sagas because of the following benefits:
- ES6 Generators make asynchronous flows easy to read, write and test (kudos to Redux Saga docs);
- I loved how the ```call()``` returns a pure object, and how I am able to test the generators with simple checks outside of the Redux;
- Sagas provide a separate place to handle all of the side effects, and actions/reducers remain as pure functions;
- Redux-saga has nice helpers, and lets me to have control over when to listen for certain actions.

Some alternatives may be:
- Using ```store.subscribe```, but in my experience it makes it less obvious where the side-effects happen, so I love the middleware approach more;
- Redux Promise Middleware setup which works well for calling RESTful APIs, and let’s you to avoid the generator syntax and some boilerplate. May be a great fit for simpler applications, but for complex ones I’d probably choose sagas;
- Redux Observable with RxJS seems like a powerful option, but I didn’t have any experience with it yet, so cannot comment.

----------
> I can see you are doing binding in constructor? What are other binding options and why you chose current one?

Yeah. It is one of the ways recommended in React Docs (https://reactjs.org/docs/faq-functions.html#how-do-i-bind-a-function-to-a-component-instance). Some alternatives would be:

- Using bind or arrow functions in render, but this approach may introduce performance issues, because the functions will be reallocated on every render. The impact is not a big deal in most cases, but when performance matters, I try to stay away from this approach;
- Using arrow functions in class properties which adopt ```this``` binding automatically. In a way it is a better approach than binding in the constructor, because it will reduce the code repetition. The only thing to consider is that it’s now in the Stage-3 proposal state, and the syntax has not been fully standardized yet. It also can have some minor implications on performance, as arrow functions in class properties get transpiled into arrow functions defined in the constructor and not in the prototype. This also removes the ability to spy on or mock class methods. It may be a premature optimization, but these are things that swayed me to stay with the constructor bindings.

----------
> You are using PrivateRoute pattern for routing? What is the reason to checkAuthToken in Servers screen? PrivateRoute is supposed to eliminate this need for explicit checking.

My bad. I did the ```checkAuthToken``` implementation before I used the ```PrivateRoute``` pattern, and then I didn’t remove the redundant ```checkAuthToken```. Will do, thanks for pointing it out :)

----------
> Is there any specific reasoning why you chose to implement functional components as const arrow functions and not just simple functions?

Mostly because they have an implicit return, and are more compact and clean syntax-wise.

----------
> What would be your approach towards app that has huge amount of available api calls/actions? In current setup each axios request is written separately? Also in current app you would have(if there is 50-100 available API calls) very bloated actionsTypes.js - how would you solve that?

In a current setup - yes, requests are written separately. If the app started growing, I’d probably start splitting up the API calls/actions into context specific parts, just to avoid having everything under the same folder or in the same file, and to make the code more manageable. As for the action types, I would probably create a separate ```actionTypes.js``` file inside the ```actions``` folder of each entity/context, and place the types there, instead of bloating a single file.
Speaking of huge amounts of API calls, I’d consider pros and cons of having GraphQL and Relay or Apollo implemented. The initial setup has a learning curve to it, but I like how easy it becomes to manage data queries that individual components need to make to the API. Also there will be no need for managing the REST API code bloat with hundreds of actions.
Another benefit I see is that one GraphQL query can call many resolvers to construct a nested response with multiple resources. It means, that Relay for e.g. will collect all the queries from individual components, compile them into a single big query, and make just one API call to the back-end over the network to get the data. There are cons to it also, but having used it in production, I would at least discuss this approach with the team.

----------
> I can see you are using named imports for your lodash functions? Could you explain why it is better over import _ from 'lodash' and then calling function _.[functionName]()?

I am importing the ```lodash.get``` as a single module to free myself from installing the whole lodash package. It’s better because:
- I’m explicitly importing only one single method which my application needs to use, and unnecessary code doesn’t get built with the application;
- By importing it as a single module I save a minor amount storage on the disk. Using the ```import get from 'lodash/get';``` would also get the job done to keep the build away from the unnecessary code, but the whole lodash package will still be installed.

----------
> In current app you need to persist only token? Lets assume you have big SPA where some data needs to be persisted and some not? How would you approach that? Also, how would you approach avoiding unnecessary http requests when needed data is already in storage? Or what would be your approach towards persisting data with "time to live"(i.e. you request user billing history and keep it in storage, but it is considered to be outdated every 30minutes)?

Yes, the current app persists only the token.
I would use the ```redux-persist``` for persistence/rehydration, and then exclude the parts of state that shouldn’t be persisted through a blacklist array in a config object.
To avoid unnecessary requests I’d implement selectors with the ```Reselect``` package, and in my sagas would do something like ```if (selector(getState())) { … }``` with an abort operation inside, so that the API doesn’t get triggered for nothing.
As for the TTL, in this project the simplest way is to persist a timestamp with the data, and then based on a timestamp and current time during the rehydration process decide whether the data needs to be abandoned. In my current project we use sessionStorage. Could be an option too.

----------
> How would you implement loading indicator for login/servers/in general any screen whit current setup of your app?

I’d use a boolean ```isLoading``` state variable which would be set to “true” right before the request, and set to ```false``` after the response has arrived. Then I’d check the ```state.isLoading``` in a render() to present the loading indicator.

----------
> Could you extend your PrivateRoute so it would be possible to pass in component that will be shown when auth requirements are not met?

Sure. I’ve added the possibility to pass an optional ```redirectComponent``` prop to the PrivateRoute.

----------
> Lets assume you have more than one PrivateRoute and user is trying to access one of them without auth token. User is redirected to login and from there to servers(which was not the one that user was trying to access via direct link), how would solve such case when after login user has to land to that PrivateRoute which he/she intended to visit in the first place?

I’ve implemented the solution for demonstration. I had it halfway implemented already. So the main idea is to pass the ```location``` property (which contains the ```pathname``` of a page user tried to visit) to the login component, so we would be able to redirect user back to that path after a successful login. I’m passing the location property inside my PrivateRoute’s redirect, which points to the login page. It looks like this:

```
    <Redirect
      to={{
        pathname: ROUTE_LOGIN,
        state: { from: props.location },
      }}
    />
```

So now on the login page, after a successful login, I am able to do:
```
    const { pathname } = this.props.location.state.from || ROUTE_SERVER_LIST;
    
    if (token) {
      this.props.push(pathname);
    }
```

And it redirects user back to the original route.
Another way to implement this in login page is to use a declarative ```<Redirect to={…} />``` inside the render method instead of the imperative “push”.

----------
> I can see babel config is done in package.json(this is the result of ejected create-react-app) - what is your thoughts on such approach? Any other possible implementation of babel config?

I usually use the ```.babelrc``` config file in the root directory for this matter. It makes ```package.json``` file less cluttered, makes for a better separation of concerns, and just seems easier to modify and find settings. But in this project ```package.json``` wasn’t very bloated, so I figured it’s okay to leave it as it is for now. The only benefit I see ```package.json``` configuration can have is the “single source of truth” approach to configs for a project. But then again, it would only work for small projects, where the file is actually manageable.

----------
Thanks.

Best regards.
